### PR TITLE
JDBC backend search filter for listing only devices or only gateways

### PR DESCRIPTION
**Pull request by gdimitropoulos-sotec**
Hono [Issue for PR](https://github.com/eclipse-hono/hono/issues/3517) 

Based on [PR](https://github.com/eclipse-hono/hono/pull/3518) 

Implement gateway filter for JDBC (H2 & Postgresql) device registry:
- add optional parameter "isGateway"
- functionality for listing only devices or gateways
- add documentation

Also-by: Matthias Feurer m.feurer@sotec.eu
Signed-off-by: g.dimitropoulos <g.dimitropoulos@sotec.eu>

### CODE REVIEW

File: site/documentation/content/api/management/device-registry-v1.yaml [hunk 1]
```diff
@@ -1771,6 +1773,20 @@ components:
             value: "{\"field\":\"/ext/brand\"}"
           sort descending:
             value: "{\"field\":\"/status/created\",\"direction\":\"desc\"}"
+      isGateway:
+        name: isGateway
+        in: query
+        description: |
+          Boolean filter parameter for searching only devices or gateways.
```
Comment (sophokles73) [inline site/documentation/content/api/management/device-registry-v1.yaml:1780]: what if I do not specify the query parameter at all?

**Review by sophokles73 — COMMENTED** [review]
Can you also please rebase the PR?

File: site/documentation/content/api/management/device-registry-v1.yaml (hunk 1 above)
Comment (gdimitropoulos-sotec) [reply site/documentation/content/api/management/device-registry-v1.yaml:1780]: Then we'll get both devices and gateways (Hono Default search functionality)

**Comment by gdimitropoulos-sotec** [comment]
The branch is rebased

File: site/documentation/content/api/management/device-registry-v1.yaml (hunk 1 above)
Comment (sophokles73) [reply site/documentation/content/api/management/device-registry-v1.yaml:1780]: My point is that IMHO we should document this explicitly ...

**Comment by sophokles73** [comment]
hmm, GitHub still seems to believe that there are conflicts with the main branch that need to be resolved ...

File: site/documentation/content/api/management/device-registry-v1.yaml (hunk 1 above)
Comment (gdimitropoulos-sotec) [reply site/documentation/content/api/management/device-registry-v1.yaml]: Done

File: core/src/main/java/org/eclipse/hono/util/RegistryManagementConstants.java [hunk 2]
```diff
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2019, 2021 Contributors to the Eclipse Foundation
+ * Copyright (c) 2019, 2021, 2023 Contributors to the Eclipse Foundation
```
Comment (sophokles73) [inline core/src/main/java/org/eclipse/hono/util/RegistryManagementConstants.java:2]: we indicate the range by means of specifying the year of initial creation and the year of the latest change. So in this case `2019, 2023`. Can you please also change this in the other classes accordingly?

File: site/documentation/content/api/management/device-registry-v1.yaml [hunk 3]
```diff
@@ -1687,7 +1688,7 @@ components:
       filterJson:
         name: filterJson
         in: query
-        description: |
+        description: |
```
Comment (sophokles73) [inline site/documentation/content/api/management/device-registry-v1.yaml:1691]: can you please remove the trailing whitespaces?

File: services/base-jdbc/src/main/java/org/eclipse/hono/service/base/jdbc/store/device/TableManagementStore.java [hunk 4]
```diff
@@ -926,25 +958,41 @@ private List<CommonCredential> parseCredentials(final ResultSet result) {
...
                                                           final SpanContext spanContext) {
 
 
-        final var filter = filters.stream().findFirst();
+        final Statement findDeviceSqlStatement;
+        final Statement countStatement;
+        final String field;
+        final String value;
+
+        if (isGateway.isPresent()) {
```
Comment (sophokles73) [inline services/base-jdbc/src/main/java/org/eclipse/hono/service/base/jdbc/store/device/TableManagementStore.java:974]: so, my understanding is that I can  search using *either* filters *or* the `isGateway` predicate, right? But I cannot search non-gateway devices having a particular property value?

File: site/documentation/content/api/management/device-registry-v1.yaml (hunk 3 above)
Comment (gdimitropoulos-sotec) [reply site/documentation/content/api/management/device-registry-v1.yaml]: Done

File: services/base-jdbc/src/main/java/org/eclipse/hono/service/base/jdbc/store/device/TableManagementStore.java [hunk 5]
```diff
@@ -913,7 +945,7 @@ private List<CommonCredential> parseCredentials(final ResultSet result) {
         final var entries = result.getRows(true);
 
         return entries.stream()
-                .map(o -> o.getString(DATA))
+                .map(o -> o.getString("data"))
```
[the remark this answers is not in the document]
Comment (gdimitropoulos-sotec) [reply services/base-jdbc/src/main/java/org/eclipse/hono/service/base/jdbc/store/device/TableManagementStore.java]: Done

File: core/src/main/java/org/eclipse/hono/util/RegistryManagementConstants.java (hunk 2 above)
Comment (gdimitropoulos-sotec) [reply core/src/main/java/org/eclipse/hono/util/RegistryManagementConstants.java]: Done

File: services/base-jdbc/src/main/java/org/eclipse/hono/service/base/jdbc/store/device/TableManagementStore.java (hunk 4 above)
Comment (gdimitropoulos-sotec) [reply services/base-jdbc/src/main/java/org/eclipse/hono/service/base/jdbc/store/device/TableManagementStore.java:974]: Yes, that is correct.

File: services/device-registry-base/src/main/java/org/eclipse/hono/deviceregistry/service/device/AbstractDeviceManagementService.java [hunk 6]
```diff
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2020, 2022 Contributors to the Eclipse Foundation
+ * Copyright (c) 2020, 2022, 2023 Contributors to the Eclipse Foundation
```
Comment (sophokles73) [inline services/device-registry-base/src/main/java/org/eclipse/hono/deviceregistry/service/device/AbstractDeviceManagementService.java:2]: can you please remove the 2022?

File: services/device-registry-base/src/main/java/org/eclipse/hono/deviceregistry/service/device/AbstractDeviceManagementService.java [hunk 7]
```diff
@@ -225,6 +225,7 @@ protected Future<Result<Void>> processDeleteDevicesOfTenant(final String tenantI
      *                   retrieve the whole result set page by page.
      * @param filters A list of filters. The filters are predicates that objects in the result set must match.
      * @param sortOptions A list of sort options. The sortOptions specify properties to sort the result set by.
+     * @param isGateway Optional filter for searching all gateways or only devices.
```
Comment (sophokles73) [inline services/device-registry-base/src/main/java/org/eclipse/hono/deviceregistry/service/device/AbstractDeviceManagementService.java:228]: can you please make it more explicit what will be returned if the given parameter is `Optional.empty()`?

File: services/device-registry-base/src/main/java/org/eclipse/hono/service/management/device/DeviceManagementService.java [hunk 8]
```diff
@@ -102,6 +102,7 @@ public interface DeviceManagementService {
      *             Implementations <em>must not</em> invoke the {@link Span#finish()} nor the {@link Span#finish(long)}
      *             methods. However,implementations may log (error) events on this span, set tags and use this span
      *             as the parent for additional spans created as part of this method's execution.
+     * @param isGateway Optional filter for searching all gateways or only devices.
```
Comment (sophokles73) [inline services/device-registry-base/src/main/java/org/eclipse/hono/service/management/device/DeviceManagementService.java:105]: can you please make it more explicit what will be returned if an empty optional is passed in?

File: services/device-registry-base/src/main/java/org/eclipse/hono/service/management/device/DelegatingDeviceManagementHttpEndpoint.java [hunk 9]
```diff
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2020, 2022 Contributors to the Eclipse Foundation
+ * Copyright (c) 2020, 2022, 2023 Contributors to the Eclipse Foundation
```
Comment (sophokles73) [inline services/device-registry-base/src/main/java/org/eclipse/hono/service/management/device/DelegatingDeviceManagementHttpEndpoint.java:2]: please remove 2022

File: services/device-registry-base/src/main/java/org/eclipse/hono/service/management/device/DelegatingDeviceManagementHttpEndpoint.java [hunk 10]
```diff
@@ -186,6 +186,12 @@ private void doSearchDevices(final RoutingContext ctx) {
                 RegistryManagementConstants.PARAM_FILTER_JSON, Filter.class);
         final Future<List<Sort>> sortOptions = decodeJsonFromRequestParameter(ctx,
                 RegistryManagementConstants.PARAM_SORT_JSON, Sort.class);
+        final Future<Optional<Boolean>> isGateway = getRequestParameter(
+                ctx,
+                RegistryManagementConstants.PARAM_IS_GATEWAY,
+                Optional.empty(),
+                CONVERTER_BOOLEAN,
+                value -> true);
```
Comment (sophokles73) [inline services/device-registry-base/src/main/java/org/eclipse/hono/service/management/device/DelegatingDeviceManagementHttpEndpoint.java:194]: you will need to include this Future in the `CompositeFuture.all()` below to prevent a potential race condition

File: services/device-registry-base/src/test/java/org/eclipse/hono/service/management/device/DelegatingDeviceManagementHttpEndpointTest.java [hunk 11]
```diff
@@ -1,5 +1,5 @@
 /**
- * Copyright (c) 2020, 2021 Contributors to the Eclipse Foundation
+ * Copyright (c) 2020, 2021, 2023 Contributors to the Eclipse Foundation
```
Comment (sophokles73) [inline services/device-registry-base/src/test/java/org/eclipse/hono/service/management/device/DelegatingDeviceManagementHttpEndpointTest.java:2]: please remove 2021

File: services/device-registry-jdbc/src/main/java/org/eclipse/hono/deviceregistry/jdbc/impl/DeviceManagementServiceImpl.java [hunk 12]
```diff
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2020, 2022 Contributors to the Eclipse Foundation
+ * Copyright (c) 2020, 2022, 2023 Contributors to the Eclipse Foundation
```
Comment (sophokles73) [inline services/device-registry-jdbc/src/main/java/org/eclipse/hono/deviceregistry/jdbc/impl/DeviceManagementServiceImpl.java:2]: please remove 2022 and make sure to check all your other touched files for the same issue...

File: service-base/src/main/java/org/eclipse/hono/service/http/AbstractHttpEndpoint.java [hunk 13]
```diff
@@ -83,6 +83,13 @@ public abstract class AbstractHttpEndpoint<T extends ServiceConfigProperties> ex
         }
     };
 
+    /**
+     * A function that tries to parse a string into an Optional boolean.
```
Comment (sophokles73) [inline service-base/src/main/java/org/eclipse/hono/service/http/AbstractHttpEndpoint.java:87]: can you please be a little more specific as to what kind of string will be passed into what kind of result?

File: service-base/src/main/java/org/eclipse/hono/service/http/AbstractHttpEndpoint.java [hunk 14]
```diff
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2016, 2022 Contributors to the Eclipse Foundation
+ * Copyright (c) 2016, 2022, 2023 Contributors to the Eclipse Foundation
```
Comment (sophokles73) [inline service-base/src/main/java/org/eclipse/hono/service/http/AbstractHttpEndpoint.java:2]: please remove the 2022

File: services/base-jdbc/src/main/resources/org/eclipse/hono/service/base/jdbc/store/device/base.postgresql.sql.yaml [hunk 15]
```diff
@@ -55,21 +55,75 @@ resolveGroups: |
...
+                  WHERE tenant_id=:tenant_id AND data ->> 'via' != '') 
+         as dr)) AS deviceCount
+
 findDevicesOfTenantWithFilter: |
-   SELECT * 
-   FROM %s 
-   WHERE 
+   SELECT *
+   FROM %s
+   WHERE
```
Comment (sophokles73) [inline services/base-jdbc/src/main/resources/org/eclipse/hono/service/base/jdbc/store/device/base.postgresql.sql.yaml:92]: does this query work for a field like `/ext/my-struct/my-field`?

**Comment by sophokles73** [comment]
@gdimitropoulos-sotec before we go on with this review: I appreciate all the effort you have put into this PR so far. However, the longer I think about the changes you have introduced, the less I am convinced that this is the right way to address the issue.

The JSON functions used in the queries in conjunction with regex based replace functions etc all indicate to me that these queries will have quite substantial execution times while Hono's protocol adapters themselves do not need any of this functionality.

I guess what I am trying to say is that I believe that without substantial changes to the underlying database schema, in particular replacing the `via` JSON array with a proper relation to another table, implementing this functionality will not be possible without compromising on performance and maintainability.

In addition, I am also no longer convinced that adding a separate query parameter `isGateway` in addition to the filterJson is the right choice. Instead, and based on the database schema change mentioned above, adding a *synthetic* field `is-gateway` in conjunction with the existing filterJson should do the trick, e.g.

```json
{
  "field": "is-gateway",
  "op": "eq",
  "value": "true"
}
```

this might require special handling when parsing the query but it should make the overall user experience more convenient.

Another approach that I can think of could be to define a *query* view on top the device and tenant tables which could produce this *synthetic* property and then run the queries against the view ...

**Comment by julian-sotec** [comment]
> @gdimitropoulos-sotec before we go on with this review: I appreciate all the effort you have put into this PR so far. However, the longer I think about the changes you have introduced, the less I am convinced that this is the right way to address the issue.
> 
> The JSON functions used in the queries in conjunction with regex based replace functions etc all indicate to me that these queries will have quite substantial execution times while Hono's protocol adapters themselves do not need any of this functionality.
> 
> I guess what I am trying to say is that I believe that without substantial changes to the underlying database schema, in particular replacing the `via` JSON array with a proper relation to another table, implementing this functionality will not be possible without compromising on performance and maintainability.
> 
> In addition, I am also no longer convinced that adding a separate query parameter `isGateway` in addition to the filterJson is the right choice. Instead, and based on the database schema change mentioned above, adding a _synthetic_ field `is-gateway` in conjunction with the existing filterJson should do the trick, e.g.
> 
> ```json
> {
>   "field": "is-gateway",
>   "op": "eq",
>   "value": "true"
> }
> ```
> 
> this might require special handling when parsing the query but it should make the overall user experience more convenient.
> 
> Another approach that I can think of could be to define a _query_ view on top the device and tenant tables which could produce this _synthetic_ property and then run the queries against the view ...
> 
> WDYT?

Hi @sophokles73: We thought of this as well and agree with you in terms of general usability. We came to the conclustion that we want to accomplish this with changing the jdbc database schema in a new issue.
However - we still think that the changes done here are valuable and improve things to the gateway feature for jdbc based use cases. 
We would therefore suggest merging this PR now and taking care about the changes and your recommendations in a new issue. 
See: #3568 
WDYT?

**Comment by sophokles73** [comment]
@gdimitropoulos-sotec what's your opinion on this?

**Comment by gdimitropoulos-sotec** [comment]
> @gdimitropoulos-sotec what's your opinion on this?

Hello @sophokles73 I agree with Julian. We have discussed this together with Julian and the team and we agree, that will be better, to have a separate table to hold the relationships between a gateway and devices (instead of the "Via" attribute).

**Comment by harism** [comment]
This implementation seems to take into consideration only the `via` attribute when determining normal and gateway devices. However there are also the `memberOf` and `viaGroups` attributes for device/gateway groups which are skipped altogether now if I understood correctly, should they count as normal or gateway devices too in this search?

**Comment by gdimitropoulos-sotec** [comment]
Hi, @harism, thank you for your comment, yes we use only the via attribute to define devices and gateways. In the  [new issue](https://github.com/eclipse-hono/hono/issues/3568), we will examine also the fields memberOf and viaGroups before designing and implementing the solution. Is It possible to link your comment to the [new issue](https://github.com/eclipse-hono/hono/issues/3568)?

**Comment by sophokles73** [comment]
@gdimitropoulos-sotec are you working on this?

**Comment by gdimitropoulos-sotec** [comment]
> @gdimitropoulos-sotec are you working on this?

Hello @sophokles73 yes I am still working on this.

File: services/device-registry-base/src/main/java/org/eclipse/hono/deviceregistry/service/device/AbstractDeviceManagementService.java (hunk 6 above)
Comment (gdimitropoulos-sotec) [reply services/device-registry-base/src/main/java/org/eclipse/hono/deviceregistry/service/device/AbstractDeviceManagementService.java]: Done

File: services/device-registry-base/src/main/java/org/eclipse/hono/deviceregistry/service/device/AbstractDeviceManagementService.java (hunk 7 above)
Comment (gdimitropoulos-sotec) [reply services/device-registry-base/src/main/java/org/eclipse/hono/deviceregistry/service/device/AbstractDeviceManagementService.java:228]: Done

File: services/device-registry-base/src/main/java/org/eclipse/hono/service/management/device/DeviceManagementService.java (hunk 8 above)
Comment (gdimitropoulos-sotec) [reply services/device-registry-base/src/main/java/org/eclipse/hono/service/management/device/DeviceManagementService.java:105]: Done

File: services/device-registry-base/src/main/java/org/eclipse/hono/service/management/device/DelegatingDeviceManagementHttpEndpoint.java (hunk 9 above)
Comment (gdimitropoulos-sotec) [reply services/device-registry-base/src/main/java/org/eclipse/hono/service/management/device/DelegatingDeviceManagementHttpEndpoint.java]: Done

File: services/device-registry-base/src/main/java/org/eclipse/hono/service/management/device/DelegatingDeviceManagementHttpEndpoint.java (hunk 10 above)
Comment (gdimitropoulos-sotec) [reply services/device-registry-base/src/main/java/org/eclipse/hono/service/management/device/DelegatingDeviceManagementHttpEndpoint.java:194]: Done

File: services/device-registry-base/src/test/java/org/eclipse/hono/service/management/device/DelegatingDeviceManagementHttpEndpointTest.java (hunk 11 above)
Comment (gdimitropoulos-sotec) [reply services/device-registry-base/src/test/java/org/eclipse/hono/service/management/device/DelegatingDeviceManagementHttpEndpointTest.java]: Done

File: services/device-registry-jdbc/src/main/java/org/eclipse/hono/deviceregistry/jdbc/impl/DeviceManagementServiceImpl.java (hunk 12 above)
Comment (gdimitropoulos-sotec) [reply services/device-registry-jdbc/src/main/java/org/eclipse/hono/deviceregistry/jdbc/impl/DeviceManagementServiceImpl.java]: Done

File: service-base/src/main/java/org/eclipse/hono/service/http/AbstractHttpEndpoint.java (hunk 13 above)
Comment (gdimitropoulos-sotec) [reply service-base/src/main/java/org/eclipse/hono/service/http/AbstractHttpEndpoint.java:87]: Done

File: service-base/src/main/java/org/eclipse/hono/service/http/AbstractHttpEndpoint.java (hunk 14 above)
Comment (gdimitropoulos-sotec) [reply service-base/src/main/java/org/eclipse/hono/service/http/AbstractHttpEndpoint.java]: Done

File: services/base-jdbc/src/main/resources/org/eclipse/hono/service/base/jdbc/store/device/base.postgresql.sql.yaml (hunk 15 above)
Comment (gdimitropoulos-sotec) [reply services/base-jdbc/src/main/resources/org/eclipse/hono/service/base/jdbc/store/device/base.postgresql.sql.yaml:92]: yes, it should.

**Comment by sophokles73** [comment]
@gdimitropoulos-sotec sorry, but it doesn't look like you have changed any code at all. Can you double check?

**Comment by gdimitropoulos-sotec** [comment]
> @gdimitropoulos-sotec sorry, but it doesn't look like you have changed any code at all. Can you double check?

Hello, I am very sorry about that. I am trying to find my commit from yesterday but I can't. I just made a new commit with all the changes.

**Comment by gdimitropoulos-sotec** [comment]
Hi @sophokles73, I fix all the issues from review comments, can this PR be merged?

File: site/documentation/content/user-guide/device-registry.md [hunk 16]
```diff
@@ -138,17 +139,21 @@ The devices in the registry can be managed using the Device Registry Management
...
 
 {{% notice info %}}
+
 The JDBC based registry implementation does not support the following features:
 
-* Registration information can be retrieved using the
-  [search devices]({{< relref "/api/management#devices/searchDevicesForTenant" >}}) operation defined by the Device
-  Registry Management API. The  *filterJson* query parameter currently only allows one filter expression per request,
+* By [search devices]({{< relref "/api/management#devices/searchDevicesForTenant" >}}) operation defined by the Device
+  Registry Management API, the *filterJson* query parameter currently only allows one filter expression per request and
```
Comment (sophokles73) [inline site/documentation/content/user-guide/device-registry.md:146]: Can you please revert this change?

File: site/documentation/content/user-guide/device-registry.md [hunk 17]
```diff
@@ -138,17 +139,21 @@ The devices in the registry can be managed using the Device Registry Management
...
   the *sortJson* query parameter is (currently) being ignored.
   The result set will always be sorted by the device Id in ascending order.
+
+The Mongo DB based registry implementation does not support the following features:
+
+* By [search devices]({{< relref "/api/management#devices/searchDevicesForTenant" >}}) operation defined by the Device
+  Registry Management API, the  *isGateway* query parameter is (currently) being ignored.
```
Comment (sophokles73) [inline site/documentation/content/user-guide/device-registry.md:153]: can you please rephrase this to match the section about the JDBC registry?

**Review by sophokles73 — CHANGES_REQUESTED** [review]

File: site/documentation/content/user-guide/device-registry.md (hunk 17 above)
Comment (gdimitropoulos-sotec) [reply site/documentation/content/user-guide/device-registry.md]: Done

File: site/documentation/content/user-guide/device-registry.md (hunk 16 above)
Comment (gdimitropoulos-sotec) [reply site/documentation/content/user-guide/device-registry.md]: Done

File: site/documentation/content/user-guide/device-registry.md [hunk 18]
```diff
@@ -57,14 +57,15 @@ The tenants in the registry can be managed using the Device Registry Management
...
 The JDBC based registry implementation does not support the following features:
 
 * Tenants can be retrieved using the [search tenants]({{< relref "/api/management#tenants/searchTenants" >}})
-  operation defined by the Device Registry Management API, but the *sortJson* query parameter is
+  operation defined by the Device Registry Management API, but the *filterJson* and *sortJson* query parameters are
   (currently) being ignored. The result set will always be sorted by the tenant Id in ascending order.
 * The *alias* and *trust-anchor-group* properties defined on a tenant are being ignored by the registry. Consequently,
   multiple tenants can not be configured to use the same trust anchor(s).
-{{% /notice %}}
+* {{% /notice %}}
```
[the remark this answers is not in the document]
Comment (gdimitropoulos-sotec) [reply site/documentation/content/user-guide/device-registry.md]: Done

File: services/device-registry-base/src/test/java/org/eclipse/hono/service/management/device/AbstractDeviceManagementSearchDevicesTest.java [hunk 19]
```diff
@@ -124,12 +125,12 @@ default void testSearchDevicesWithMultipleFiltersSucceeds(final VertxTestContext
...
-                            List.of(), NoopSpan.INSTANCE))
-            .onComplete(ctx.succeeding(s -> {
-                ctx.verify(() -> {
-                    assertThat(s.getStatus()).isEqualTo(HttpURLConnection.HTTP_OK);
+                        .compose(ok -> getDeviceManagementService()
+                                .searchDevices(tenantId, pageSize, pageOffset, List.of(filter1, filter2, filter3),
+                                        List.of(), Optional.empty(), NoopSpan.INSTANCE))
+                                .onComplete(ctx.succeeding(s -> {
+                                    ctx.verify(() -> {
+                                        assertThat(s.getStatus()).isEqualTo(HttpURLConnection.HTTP_OK);
```
[the remark this answers is not in the document]
Comment (gdimitropoulos-sotec) [reply services/device-registry-base/src/test/java/org/eclipse/hono/service/management/device/AbstractDeviceManagementSearchDevicesTest.java]: Done

**Comment by gdimitropoulos-sotec** [comment]
Hi @sophokles73  thank you for your comments.  The "isGateway" is documented in both the JavaDoc and device-registry-v1.yaml file.

File: services/base-jdbc/src/main/java/org/eclipse/hono/service/base/jdbc/store/device/TableManagementStore.java (hunk 4 above)
Comment (sophokles73) [reply services/base-jdbc/src/main/java/org/eclipse/hono/service/base/jdbc/store/device/TableManagementStore.java:974]: Still neither the JavaDoc nor device-registry-v1.yaml mention that you can either use filter or isGateway, but not both ...

File: services/device-registry-base/src/test/java/org/eclipse/hono/service/management/device/AbstractDeviceManagementSearchDevicesTest.java [hunk 20]
```diff
@@ -124,12 +125,12 @@ default void testSearchDevicesWithMultipleFiltersSucceeds(final VertxTestContext
...
-                            List.of(), NoopSpan.INSTANCE))
-            .onComplete(ctx.succeeding(s -> {
-                ctx.verify(() -> {
-                    assertThat(s.getStatus()).isEqualTo(HttpURLConnection.HTTP_OK);
+                        .compose(ok -> getDeviceManagementService()
+                        .searchDevices(tenantId, pageSize, pageOffset, List.of(filter1, filter2, filter3),
+                                List.of(), Optional.empty(), NoopSpan.INSTANCE))
+                        .onComplete(ctx.succeeding(s -> {
+                            ctx.verify(() -> {
+                                assertThat(s.getStatus()).isEqualTo(HttpURLConnection.HTTP_OK);
```
Comment (sophokles73) [inline services/device-registry-base/src/test/java/org/eclipse/hono/service/management/device/AbstractDeviceManagementSearchDevicesTest.java:133]: can you please undo the indentation?

**Review by sophokles73 — CHANGES_REQUESTED** [review]

File: services/device-registry-base/src/test/java/org/eclipse/hono/service/management/device/AbstractDeviceManagementSearchDevicesTest.java (hunk 20 above)
Comment (gdimitropoulos-sotec) [reply services/device-registry-base/src/test/java/org/eclipse/hono/service/management/device/AbstractDeviceManagementSearchDevicesTest.java]: Done

**Comment by gdimitropoulos-sotec** [comment]
Hi @sophokles73, I understand if the isGateway is documented. I changed the documentation so that is clear if the parameter  isGateway is used then the parameter  filter will be ignored.

**Review by sophokles73 — APPROVED** [review]
LGTM
